---
title: "Weighted Gene Correlation Network Analysis for Larval CASE RNAseq"
author: "Megan Guidry"
date: "03/03/2023"
output: github_document
---

Load packages
```{r, results="hide"}
library(WGCNA) # installed this package with `BiocManager::install("WGCNA")`
library(dplyr)
library(zoo)
library(DESeq2)
# for heatmap
# library(devtools)
# install_github("jokergoo/ComplexHeatmap") first run these - commented out to avoid running a second time...
library(ComplexHeatmap)
library(circlize)
library(reshape)
library(ggplot2)
library(hrbrthemes)
```




Set working directory
```{r}
setwd("/home/mguidry/repos/Larval-Oyster-CASE-RNA/2023_WGCNA_MG")
getwd()
```

Load data - Raw count matrix file and treatment/metadata file. 
This is the same files that went into DESeq. You could start with the filtered count file that you wrote out in the DESeq script, but we'll just do that same filtering of the raw counts below.
```{r}
#count matrix with rownames=sample ID and colnames=gene ID
##these are my raw counts, will filter in next steps
count_data <- as.data.frame(read.csv("gene_count_matrix.csv", row.names="gene_id"))
dim(count_data)

#metadata about the treatments
treatment_data <- read.csv(file="treatment_data.csv", sep=',', header=TRUE)
```

The following setting is important, do not omit. (as recommended by WGCNA authors - view tutorial)
```{r}
options(stringsAsFactors = FALSE)
```


Preprocessing the data
Preliminary filtering of the counts with P over A, Meaning that I wanted to remove all rows that have less than 21.4% of the samples with less than 5 counts. 21.4% was chosen because that is 3/14, or the minimum ratio of samples per one treatment. I am starting with the raw counts ('count_data' variable).

[Read here for the pre processing steps using WGCNA!](https://bmcbioinformatics.biomedcentral.com/track/pdf/10.1186/1471-2105-9-559.pdf)
```{r}
#Filtering the count data
#set filter values for PoverA, P percent of the samples have counts over A
filt <- filterfun(pOverA(0.214,5))

#create filter for the counts data
tfil <- genefilter(count_data, filt)

#identify transcripts to keep by count filter
keep <- count_data[tfil,]

#identify transcript list
gn.keep <- rownames(keep)

#data filtered in PoverA, P percent of the samples have counts over A
count_data_filt <- as.data.frame(count_data[which(rownames(count_data) %in% gn.keep),])
dim(count_data) #62745 by 14
dim(count_data_filt) #29338 by 14
head(count_data_filt,10)

# sanity check - make sure all of your samples in both datasets
dim(treatment_data)[1] ==  dim(count_data_filt)[2]
# TRUE = count data and treatment data matrices contain all 14 samples - yay
```

The question of how to normalize this dataset came up. The creators of WGCNA insist that "properly normalized" data is required. It turns out transforming your data (ex vst or log) is a type of normalization method. Going forward in this script are two different normalization methods. One use the TMM method from EdgeR and then log2 transforms it from there. As far as I can tell, the reason to log2 (or any base log) transform any data is to make it more symetrical. It is likely true that my data has a skew in it.

The other method is to use the variance stabilizing transformation on just the raw counts. "Many common statistical methods for exploratory analysis of multidimensional data, especially methods for clustering and ordination (e.g., principal-component analysis and the like), work best for (at least approximately) homoskedastic data; this means that the variance of an observable quantity (i.e., here, the expression strength of a gene) does not depend on the mean. In RNA-Seq data, however, variance grows with the mean." <https://www.biostars.org/p/109104/> <http://www.nxn.se/valent/2017/10/15/variance-stabilizing-scrna-seq-counts> The count data is in no way normally distributed "The process of counting implies that variation will propagate as the number of events increase. The effect of this is that there will be an inherent relation between mean (expected value) and variance of counts" From what I can gather the vst transformation removes the relationship between the mean and the variance.

**We will use a variance stabilizing transformation here (vst transformation)**
This transformation is recommended in the R documentation for using WGCNA on RNAseq data <https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html/>


Create 'dds' object (DESeqDataSet) & vst transform the 'dds' object for WGCNA 
Follow the parameters used in previous DESeq analysis -- See `MG-DESEQ2/DESeq2.md` Matrix building section
Here, as in the DESeq2 script, I am factoring the data by treatment and library effects.
Then we'll transform the data and select only the top 5000 most variable genes to look at in the analysis
```{r}
dds1 <- DESeqDataSetFromMatrix(countData = count_data_filt,
                              colData = treatment_data,
                              design = ~ treatment ) # column name of the treatment information as the design 
dds1
# Tell the matrix what the factors in the treatment variable are and level the matrix by the treatment
dds1$treatment <- factor(dds1$treatment, levels = c("CON","CA", "SE", "CASE"))
levels(dds1$treatment)
dds1

dds2 <- dds1
dds2$library <- factor(dds2$library, levels = c("one","two", "three", "four"))
levels(dds2$library) # change the levels to a different variable
design(dds2) <- formula(~ library + treatment) # remake design with library first
dds2 # view the DESeqDataSet - notice the colData containng our critical treatment and sample ID data, rownames, etc. 



# transform the data  ========================================================== #
# run in order (kept name as dds2_vst)
dds2_vst <- vst(dds2) # transform it vst (variance stabilized transformation)
dds2_vst <- assay(dds2_vst) # call only the transformed counts in the dds object

# lets pair down to look at the top 5,000 most variable genes instead of the list of 29,338 genes
MostVaryGenes <- head(order(rowVars(dds2_vst), decreasing = TRUE), 5000)
# make that information into a matrix
VMostVaryGenes_Mat <- dds2_vst[ MostVaryGenes, ] #which in the DESeq2 style matrix are those genes?
dim(VMostVaryGenes_Mat)
# make that matrix a dataframe
dds2_vst_5000 <- as.data.frame(VMostVaryGenes_Mat)
dim(dds2_vst_5000)
dds4<- t(dds2_vst_5000) # transpose columns to rows and vice versa
dim(dds4)
```


WGCNA Sample tree - omit outlier sample(s)
```{r}
# checks before we start....
dim(dds4) #  14 samples; 5000 genes
gsg = goodSamplesGenes(dds4, verbose = 3);  # We first check for genes and samples with too many missing values:
gsg$allOK # If the statement returns TRUE, all genes have passed the cuts. 
```

Determine outlier sample(s) with cluster tree
```{r}
# call the cluster and set window dimensions to view..
sampleTree = hclust(dist(dds4), method = "average") # Next we cluster the samples (in contrast to clustering genes that will come later)  to see if there are any obvious outliers.
sizeGrWindow(12,9) # The user should change the dimensions if the window is too large or too small.
par(cex = 0.6);
par(mar = c(0,4,2,0))
# output the tree as .png file..
png("outputs/sampleClusterTree_Precut.png", 1000, 1000, pointsize=20)
plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5, 
     cex.axis = 1.5, cex.main = 2) 
#abline(h = , col = "red") # add line to plot to show the cut-off outlier samples 
dev.off()
```

Check out your tree! I will not be cutting any samples, based on my tree since there are no obvious outliers.
NOTE: you can insert a horizontal line (using the `abline()` function in the code above) demonstrating where we likely want to cut our tree to omit outliers. 

Cutting the outlier(s) and re-plotting the tree. 
I did not use these steps for these data given the lack of outliers. But in the event you have an outlier, you'd want to follow the code below to obtain a new data matrix sans outliers and make a new post-cut tree.
```{r}
# # cut the tree and omit  ========================================================== #
# clust = cutreeStatic(sampleTree, cutHeight = 80, minSize = 10) # Determine cluster under the line
# table(clust) # 0 = cut; 1 = kept; says it will cut 1 and save 35; exactly what we want!  
# keepSamples = (clust==1) # 'keepsamples' boolean to call the main dataset - notice there is one occurrence of FLASE - this is sample SG59
# 
# # integrate keepsamples  ========================================================== #
# dds4 = dds4[keepSamples, ] # integreat the boolean 'keepsamples' to ommit oultilers determined in the sample tree above 
# nGenes = ncol(dds4) # number of genes == 29338 
# nSamples = nrow(dds4) # number of samples == 35  - the cut tree removed 1 sample 
# 
# # plot the tree with the 'keep samples'  =========================================== #
# sampleTree2 = hclust(dist(dds4), method = "average") # Next we cluster the samples (in contrast to clustering genes that will come later)  to see if there are any obvious outliers.
# png("outputs/ClusterTree_Postcut.png", 1000, 1000, pointsize=20)
# plot(sampleTree2, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5, 
#      cex.axis = 1.5, cex.main = 2)
# dev.off()
```

Updating matrices for following analyses (do this whether or not you cut samples)
```{r}
dim(dds4)
dim(treatment_data)

# Form a data frame analogous to expression data that will hold the clinical traits.
samples = rownames(dds4);# start new variable 'd7.Samples' calling the row names of the gene data (sample as 'Mouse' in trait data)
TreatRows = match(samples, treatment_data$sample); # match the names - calls the list number of 'd7.Samples' matching 'd7.Treatment.data$Sample.Name'
Traits = treatment_data[TreatRows, -1]; # insert TreatRows as the row numbers in 'treatment_data'
rownames(Traits) = treatment_data[TreatRows, 1]; # inserts the new TreatRows - matches sample IDs
all(rownames(Traits) == rownames(dds4))  # should be TRUE
dim(Traits)
```

Prepare Trait data (CA and SE groupings)
```{r}
# Coastal acidification groups ===================================================== #
CA_traits     <-  Traits %>% dplyr::select('pCO2') %>% # pCO2 treatment as coastal acidification (CA) vs. ambient (amb)
                            dplyr::mutate(CA = as.factor(as.numeric(pCO2 == "CA")))  %>%  # call occurrence of 'CA' as 0s and 1s (factor)
                            dplyr::mutate(amb = as.factor(as.numeric(pCO2 == "ambient")))  %>%  # call occurrence of 'amb'  as 0s and 1s (factor)
                            dplyr::select(-pCO2)
CA_traits  # final dataset of 0,1 for treatment groups - CA only!

# Sewage effluent groups
SE_traits     <-  Traits %>% dplyr::select('effluent') %>% # effluent as sewage effluent(SE) vs. ambient (amb)
  dplyr::mutate(SE = as.factor(as.numeric(effluent == "SE")))  %>%  # call occurrence of 'A' as 0s and 1s (factor)
  dplyr::mutate(amb = as.factor(as.numeric(effluent == "ambient")))  %>%  # call occurrence of 'M'  as 0s and 1s (factor)
  dplyr::select(-effluent)
SE_traits  # final dataset of 0,1 for treatment groups - SE only!

# Combined treatments
Traits.CASE <-   Traits %>% dplyr::select('treatment') %>% # combined treatments as CASE vs. CA vs. SE vs. CON
  dplyr::mutate(CASE = as.factor(as.numeric(treatment == "CASE")))  %>%  # call occurrence of 'CASE' as 0s and 1s (factor)
  dplyr::mutate(CA = as.factor(as.numeric(treatment == "CA")))  %>%  # call occurrence of 'CA' as 0s and 1s (factor)
  dplyr::mutate(SE = as.factor(as.numeric(treatment == "SE")))  %>%  # call occurrence of 'SE' as 0s and 1s (factor)
  dplyr::mutate(CON = as.factor(as.numeric(treatment == "CON")))  %>%  # call occurrence of 'CON' as 0s and 1s (factor)
  dplyr::select(-treatment)
Traits.CASE
```

Cluster samples by Trait
*make sure you are using the right sample tree!!*
Since I didn't cut any samples, I am using the original tree built.
```{r}
# CA treatment
png("outputs/ClusterTree_CA.png", 1000, 1000, pointsize=20)
traitColors_CA = labels2colors(CA_traits); # Convert traits to a color representation: white means low, red means high, grey means missing entry
plotDendroAndColors(sampleTree, traitColors_CA, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(CA_traits), 
                    main = "Sample dendrogram and trait heatmap")
dev.off()

# SE treatment 
png("outputs/ClusterTree_SE.png", 1000, 1000, pointsize=20)
traitColors_SE = labels2colors(SE_traits); # Convert traits to a color representation: white means low, red means high, grey means missing entry
plotDendroAndColors(sampleTree, traitColors_SE, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(SE_traits), 
                    main = "Sample dendrogram and trait heatmap")
dev.off()

# combined treatment
png("outputs/ClusterTree_combined.png", 1000, 1000, pointsize=20)
traitColorsGroup = labels2colors(Traits.CASE); # Convert traits to a color representation: white means low, red means high, grey means missing entry
plotDendroAndColors(sampleTree, traitColorsGroup, # Plot the sample dendrogram and the colors underneath.
                    groupLabels = names(Traits.CASE), 
                    main = "Sample dendrogram and trait heatmap")
dev.off()


# write some of the data thus far.... 
#save(dds4, Traits, file = "data/dataInput.RData")
#write.csv(dds4, "data/vstTransformed_WGCNAdata.csv") # # write the vst transformed data 
```


Setting soft threshold

The next thing to do is determine the power or soft threshold to apply for the adjacency matrix. This is because we want to make a weighted matrix, where each gene has a weighted connection, not a hard threshold of either connected or not. From my understanding, by raising connection/adjacencies to a power there in lies the weighted-ness of their connection.

I need to pick a power where scale free topolgy of the matrix is met. From some googling, this is similar to a scale free network. "A scale-free network can be constructed by progressively adding nodes to an existing network and introducing links to existing nodes with preferential attachment so that the probability of linking to a given node i is proportional to the number of existing links k\_i that node has" So there is a higher chance of being connected to a node gene if it has more connections. That makes intuitive sense, which is why it has to reach this criterion to be passable for further analysis. <https://mathworld.wolfram.com/Scale-FreeNetwork.html>

"If the scale-free topology fit index fails to reach values above 0.8 for reasonable powers (less than 15 for unsigned or signed hybrid networks, and less than 30 for signed networks) and the mean connectivity remains relatively high (in the hundreds or above), chances are that the data exhibit a strong driver that makes a subset of the samples globally different from the rest. The difference causes high correlation among large groups of genes which invalidates the assumption of the scale-free topology approximation."<https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html>

"If the lack of scale-free topology fit turns out to be caused by an interesting biological variable that one does not want to remove (i.e., adjust the data for), the appropriate soft-thresholding power can be chosen based on the number of samples as in the table below. This table has been updated in December 2017 to make the resulting networks conservative." For a signed network with less than 15 samples, 18 is an appropreate power to use.

<https://peterlangfelder.com/2018/11/25/signed-or-unsigned-which-network-type-is-preferable/>
```{r}
dim(dds4) # again double check you have the correct data...
# Choose a set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12, to=20, by=2))
# Call the network topology analysis function
sft = pickSoftThreshold(dds4, powerVector = powers, verbose = 5) #...wait for this to finish (~4mins)
#  pickSoftThreshold 
#  performs the analysis of network topology and aids the
#  user in choosing a proper soft-thresholding power.
#  The user chooses a set of candidate powers (the function provides suitable default values)
#  function returns a set of network indices that should be inspected

# png to output 
sizeGrWindow(9, 5) # set window size 
png("outputs/ScaleTopology_SoftThresh.png", 1000, 1000, pointsize=20)
        par(mfrow = c(1,2));
        cex1 = 0.9;
        
        plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], # Scale-free topology fit index as a function of the soft-thresholding power
             xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
             main = paste("Scale independence"));
        text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
             labels=powers,cex=cex1,col="red");
        abline(h=0.90,col="red") # look at at cut off at power of 3 - this line corresponds to using an R^2 cut-off of h
        
        plot(sft$fitIndices[,1], sft$fitIndices[,5], # Mean connectivity as a function of the soft-thresholding power
             xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
             main = paste("Mean connectivity"))
        text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
dev.off() # output 


# The left panel... shows the scale-free fit index (y-axis) 
# as a function of the soft-thresholding power (x-axis). 
# We choose the power 9, which is the lowest power for which the scale-free 
# topology index curve attens out upon reaching a high value (in this case, roughly 0.90).
# The right panel.... displays the mean connectivity
# (degree, y-axis) as a function of the soft-thresholding power (x-axis).
```

Output from `pickSoftThreshold()`:
$powerEstimate
[1] 8

$fitIndices
   Power  SFT.R.sq       slope truncated.R.sq    mean.k.  median.k.    max.k.
1      1 0.6553952  0.91777617      0.5583942 2061.17651 2237.17663 3020.9580
2      2 0.0102838 -0.04070285     -0.1131503 1189.60412 1269.90770 2218.5634
3      3 0.4076685 -0.31086408      0.2710878  793.88928  800.50941 1756.6484
4      4 0.5816715 -0.44604552      0.4644515  575.11022  538.66875 1447.1283
5      5 0.7003088 -0.52325360      0.6162835  439.27585  376.76377 1224.7717
6      6 0.7724233 -0.57515475      0.7124488  348.19155  272.97570 1058.9239
7      7 0.8300810 -0.61511936      0.7879034  283.66356  202.17005  928.5048
8      8 0.8755421 -0.64574132      0.8460359  236.02331  153.22324  823.7731
9      9 0.8784803 -0.68391620      0.8579557  199.70247  117.12063  744.5742
10    10 0.8828040 -0.71468779      0.8684172  171.28970   91.02291  677.7473
11    12 0.8931298 -0.77985851      0.8912323  130.13360   57.71002  573.6367
12    14 0.8934559 -0.83045375      0.9034531  102.17296   37.97835  497.5125
13    16 0.8979320 -0.87186899      0.9230659   82.23779   25.75770  437.7996
14  **18 0.9158328** -0.90094639      0.9468291   67.49388   18.21410  389.1182
15    20 0.9001673 -0.95426942      0.9397137   56.27108   12.99427  350.9278

I'll pick 18 as my softpower threshold because this is where the Scale Free Topology R sq value is >=0.9. We also have less than 15 samples in this dataset which fits the parameters given [in the R documentation for WGCNA](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html/) for choosing a soft power threshold of 18 for signed networks.

Start the step-wise module construction:  
Step 1 = create adjacency matrix 
https://peterlangfelder.com/2018/11/25/signed-or-unsigned-which-network-type-is-preferable/
https://www.rdocumentation.org/packages/WGCNA/10cpm/versions/1.69/topics/adjacency
https://ramellose.github.io/networktutorials/wgcna.html
https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/10cpm/TechnicalReports/signedTOM.pdf

```{r}
softPower = 18 # set your soft threshold based on the plots above 

# signed - must call the type, defaults to unsigned
adjacency_sign = adjacency(dds4, power = softPower, type="signed") # this takes a long time.. just wait...(~seconds)
```

Step 2: Turn adjacency into topological overlap matrix
Calculation of the topological overlap matrix, (TOM)
and the corresponding dissimilarity, from a given adjacency matrix.

# signed matrix
```{r}
TOM_sign       = TOMsimilarity(adjacency_sign, TOMType="signed")  # this takes a long time.. just wait... (?)
dissTOM_sign   = 1-TOM_sign
```

Step 3:Call the hierarchical clustering function - plot the tree

```{r}
#Call the hierarchical clustering function
geneTree_sign   = hclust(as.dist(dissTOM_sign), method = "average");

# Plot the resulting clustering tree (dendrogram)
sizeGrWindow(12,9)
plot(geneTree_sign, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity - SIGNED",
     labels = FALSE, hang = 0.04);
```

Step 4: Set module size and 'cutreeDynamic' to create clusters 

```{r}
# We like large modules, so we set the minimum module size relatively high:
minModuleSize = 100; # set this for the subseqent call - WGCNA authors recommend diligence when calling module size to avoid too many/too few modules...
# Module identification using dynamic tree cut:
dynamicMods_sign = cutreeDynamic(dendro = geneTree_sign, distM = dissTOM_sign,
                                 deepSplit = 1, pamRespectsDendro = FALSE,
                                 minClusterSize = minModuleSize);
table(dynamicMods_sign) # view the number of genes per module 
```

Step 5: convert numeric network to colors and plot the dendrogram
```{r}
# Convert numeric labels into colors
dynamicColors_sign = labels2colors(dynamicMods_sign) # add colors to module labels (previously numbers)
table(dynamicColors_sign) # lets look at this table...

# Plot the dendrogram and colors underneath
sizeGrWindow(8,6)
plotDendroAndColors(geneTree_sign, dynamicColors_sign, "Dynamic Tree Cut - SIGNED",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene dendrogram and module colors 'SIGNED'")
```

Step 6: Calculate Eigengenes - view their connectivity based on 'MEDiss = 1-cor(MEs)'
```{r}
# Calculate eigengenes ========================================================== # 

# MEList = moduleEigengenes(dds4, colors = dynamicColors)
MEList = moduleEigengenes(dds4, colors = dynamicColors_sign)
MEs    = MEList$eigengenes # you can view MEs, condenses gene counts down to a single number for each sample representative of that expression pattern 

# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs);
METree = hclust(as.dist(MEDiss), method = "average") # Cluster module eigengenes

# Plot the result ================================================================ #
sizeGrWindow(7, 6)
png("outputs/ClusterEigengenes.png", 1000, 1000, pointsize=20)
plot(METree, main = "Clustering of module eigengenes - SIGNED (dissimilarity calc = MEDiss = 1-cor(MEs))",
     xlab = "", sub = "")
dev.off()
```

Step 7: Specify the cut line for the dendrogram (module) - Calc MODULE EIGENGENES (mergeMEs)
not necessary for this dataset - METree above looks to be parsed well into the 9 modules.
*If you are skipping this step, move on to Step 9*


Step 8: Plot dendrogram with the cut line 'MEDissThres' 
use 'mergedColors' we needed to merge related eigengene modules together 
```{r}
# sizeGrWindow(12, 9)
# png("Analysis/Output/WGCNA/10cpm/Day7/Day7_ClusterDendrogram_signed.png", 1000, 1000, pointsize=20)
# plotDendroAndColors(geneTree_sign, cbind(dynamicColors_sign, mergedColors),
#                     c("Dynamic Tree Cut", "Merged dynamic"),
#                     dendroLabels = FALSE, hang = 0.03,
#                     addGuide = TRUE, guideHang = 0.05)
# dev.off()
```

Step 9: Commit to mergedcolors as 'MEs' and 'moduleColors'
```{r}
# Rename to moduleColors
moduleColors = dynamicColors_sign
# Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels = match(moduleColors, colorOrder)-1;

# Save module colors and labels for use in subsequent parts
#save(MEs, moduleLabels, moduleColors, file = "data/networkConstruction-stepByStep.RData")
# write csv - save the module eigengenes
#write.csv(MEs, file = "data/WGCNA_ModuleEigengenes.csv")
```

Prepare for  module trait associations - Eigengene calc - trait data as numeric
Identify modules that are significantly associated with the measured  traits (here as treatment history)

Since we already have a summary profile (eigengene) for each module,
we simply correlate eigengenes with external traits and look for the  significant associations:
```{r}
# Define numbers of genes and samples
nGenes = ncol(dds4); 
nSamples = nrow(dds4); 
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(dds4, moduleColors)$eigengenes
MEs = orderMEs(MEs0) # reorders the columns (colors/modules)
```

Module trait correlation
```{r}
# ALL TRAIT DATA

dim(Traits)  # 14  5
dim(MEs)  # 14  9
# moduleTraitCor = cor(MEs, d7.Traits, use = "p");
# moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples);

# CA 
CA_traits_asnum    <- data.frame(lapply(CA_traits, function(x) as.numeric(as.character(x))),
                                check.names=F, row.names = row.names(CA_traits))
moduleTraitCor_CA      = cor(MEs, CA_traits_asnum, use = "p");
moduleTraitPvalue_CA   = corPvalueStudent(moduleTraitCor_CA, nSamples);
print("CA Module Correlation p-values")
moduleTraitPvalue_CA


# SE
SE_traits_asnum  <- data.frame(lapply(SE_traits, function(x) as.numeric(as.character(x))),
                                       check.names=F, row.names = row.names(SE_traits))
moduleTraitCor_SE    = cor(MEs, SE_traits_asnum, use = "p");
moduleTraitPvalue_SE = corPvalueStudent(moduleTraitCor_SE, nSamples);
print("SE Module Correlation p-values")
moduleTraitPvalue_SE


# Group
Traits.CASE_asnum      <- data.frame(lapply(Traits.CASE, function(x) as.numeric(as.character(x))),
                                        check.names=F, row.names = row.names(Traits.CASE))
moduleTraitCor_CASE        = cor(MEs, Traits.CASE_asnum, use = "p");
moduleTraitPvalue_CASE     = corPvalueStudent(moduleTraitCor_CASE, nSamples);
print("CASE Module Correlation p-values")
moduleTraitPvalue_CASE
```

```{r}
dds5 = as.data.frame((dds4))
dim(dds5) 
module_grey_genes <- names(dds5)[moduleColors=="grey"]

write.csv(module_grey_genes, "Module_grey_GO/module_grey_genes.csv", quote=F, row.names = FALSE)

## make sure the csv is just a list of genes (no row or column names)
## and use it in GO-term-prep.md
```

Hub genes -- this did not do what I wanted it to...
```{r}
# CASE = as.data.frame(Traits.CASE$CASE);
# names(CASE) = "combinedCASE"
# modNames = substring(names(MEs), 3)
# ModuleMembership = as.data.frame(cor(dds4, MEs, use = "p"));
# MMPvalue = as.data.frame(corPvalueStudent(as.matrix(ModuleMembership), nSamples));
# names(ModuleMembership) = paste("MM", modNames, sep="");
# names(MMPvalue) = paste("p.MM", modNames, sep="");
# TraitSignificance = as.data.frame(cor(dds4, CASE, use = "p"));
# GSPvalue = as.data.frame(corPvalueStudent(as.matrix(TraitSignificance), nSamples));
# names(TraitSignificance) = paste("GS.", names(CASE), sep="");
# names(GSPvalue) = paste("p.GS.", names(CASE), sep="");
# module = "grey"
# column = match(module, modNames);
# moduleTaxa = moduleColors==module;
# sizeGrWindow(7, 7);
# par(mfrow = c(1,1));
# verboseScatterplot(abs(ModuleMembership[moduleTaxa, column]),
#                    abs(TraitSignificance[moduleTaxa, 1]),
#                    xlab = paste("Module Membership in", module, "module"),
#                    ylab = "Significance for CASE",
#                    main = paste("Module membership vs. significance\n"),
#                    cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)
# 
# 
# 
# 
# #ModuleMembership = as.data.frame(cor(dds4, MEs, use = "p"))
# #TraitSignificance = as.data.frame(cor(dds4, Traits.CASE_asnum, use = "p"))
# #MMPvalue = as.data.frame(corPvalueStudent(as.matrix(ModuleMembership), nSamples));
# 
# hubs    = chooseTopHubInEachModule(dds4, moduleColors)
# head(MMPvalue) 
# head(ModuleMembership)
# hg_df <- cbind(ModuleMembership,MMPvalue,TraitSignificance)
# dim(hg_df)
# head(hg_df)
# 
# grey_CASE <- hg_df[c(abs(hg_df$MMgrey) >= 0.7 & hg_df$p.MMgrey <= 0.05),c(19)]
# dim(grey_CASE)
# View(grey_CASE)
# 
# grey_CASE2 <- hg_df[moduleTaxa,c(19)]
# dim(grey_CASE2)
# View(grey_CASE2)
# 
# 
# dim(grey_CASE)
# WGCNA_hub <- as.data.frame(dds4)
# WGCNA_hub2 <- merge(WGCNA_hub,grey_CASE,by="row.names")
# head(WGCNA_hub2)
# 
# row.names(WGCNA_hub2) <- WGCNA_hub2$Row.names  
# row.names(WGCNA_hub3) <- WGCNA_hub3$Row.names  
# WGCNA_hub2$Row.names = NULL
# WGCNA_hub3$Row.names = NULL
# dim(WGCNA_hub2)
# #[1] 123 298
# dim(WGCNA_hub3)
# #[1] 16 298
# 
# sigtab_black = WGCNA_hub2
# sigtab_black2 = cbind(as(sigtab_black, "data.frame"), as(tax_table(genusglom)[rownames(sigtab_black), ], "matrix"))
# head(sigtab_black2)
# #colnames(sigtab_black2)[2] <- c("log2FoldChange_outplant")
# #colnames(sigtab_black2)[7] <- c("WGCNA_ModuleMembership")
# #colnames(sigtab_black2)[8] <- c("WGCNA_ModulePvalue")
# head(sigtab_black2)
# write.csv(sigtab_black2,"ModuleBlack_output.csv",quote=T)
# View(sigtab_black2)
# 
# 
# sigtab_black2
# head(Metadata)
# Metadata2 <- Metadata
# row.names(Metadata2) <- Metadata2$Sample_ID
# sigtab_black3 <- t(sigtab_black2)
# head(sigtab_black3)
# head(Metadata2)
# sigtab_black4 <- merge(Metadata2,sigtab_black3,by="row.names")
# sigtab_black4 <- t(sigtab_black4)
# dim(sigtab_black4)
# 
# write.csv(sigtab_black4,"module_black_all.csv",quote=F)
```


Heatmaps-- WOOHOO!

Coastal acidification (CA) heatmap
```{r}
sizeGrWindow(10,10)
# Will display correlations and their p-values
CA.matrix <-  paste(signif(moduleTraitCor_CA, 2), "\n(",
                                      signif(moduleTraitPvalue_CA, 1), ")", sep = "")
#dim(textMatrix) == dim(moduleTraitCor_treatonly)
par(mar = c(8, 9.5, 5, 3));
# Display the correlation values within a heatmap plot
png("outputs/CA_heatmap.png", 500, 1000, pointsize=20)
labeledHeatmap(Matrix = moduleTraitCor_CA,
               xLabels = names(CA_traits),
               yLabels = names(MEs),
               ySymbols = names(MEs),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = CA.matrix,
               setStdMargins = FALSE,
               cex.text = 1,
               zlim = c(-0.6,0.6),
               main = paste("Module-trait relationships - CA"))
dev.off()
```

```{r}
# this heatmap looks better
CA.text <-  as.matrix(signif(moduleTraitPvalue_CA, 3))
pa = cluster::pam(CA.text, k = 1)
col_fun = colorRamp2(c(-0.5, 0, 0.5), c("blue", "white", "red"))
pdf("outputs/CA_heatmap2.pdf", width=5, height=6)
Heatmap(moduleTraitCor_CA, 
        name = "gene_cor", 
        rect_gp = gpar(col = "grey", lwd = 1),
        column_title = "WGCNA CA treatment", 
        column_title_gp = gpar(fontsize = 12, fontface = "bold"),
        # row_title = "WGCNA modules",
        #row_km = 4, 
        column_km = 1,
        row_split = paste0("clstr", pa$clustering),
        row_gap = unit(5, "mm"),
        column_gap = unit(5, "mm"),
        # grid.text(matrix(textMatrix)),
        # border = TRUE,
        border = TRUE,
        col = col_fun,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1f", CA.text[i, j]), x, y, gp = gpar(fontsize = 10))
        })
dev.off()
```




Sewage effluent heatmap
```{r}
sizeGrWindow(10,10)
# Will display correlations and their p-values
SE.matrix <-  paste(signif(moduleTraitCor_SE, 2), "\n(",
                                     signif(moduleTraitPvalue_SE, 3), ")", sep = "")
par(mar = c(8, 9.5, 5, 3));
# Display the correlation values within a heatmap plot
png("outputs/SE_heatmap.png", 500, 1000, pointsize=20)
labeledHeatmap(Matrix = moduleTraitCor_SE,
               xLabels = names(SE_traits),
               yLabels = names(MEs),
               ySymbols = names(MEs),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = SE.matrix,
               setStdMargins = FALSE,
               cex.text = 0.8,
               zlim = c(-0.6,0.6),
               main = paste("Module-trait relationships - Treatment SE"))
dev.off()
```
SE heatmap 2
```{r}
# this heatmap looks better
d7.SECONDtreatment_text<-  as.matrix(signif(moduleTraitPvalue_Secondary, 4))
d7.SECONDtreatment_cor <-  as.matrix(signif(moduleTraitCor_Secondary, 4))
pa = cluster::pam(d7.SECONDtreatment_cor, k = 3)
col_fun = colorRamp2(c(-0.5, 0, 0.5), c("blue", "white", "red"))

pdf("/home/mguidry/wgcna_exercise/Day7_Treatments_Second_heatmap.pdf", width=5, height=6)
Heatmap(moduleTraitCor_Secondary, 
        name = "gene_cor", 
        rect_gp = gpar(col = "grey", lwd = 1),
        column_title = "Day 7 WGCNA - Treatment Groups", 
        column_title_gp = gpar(fontsize = 12, fontface = "bold"),
        # row_title = "WGCNA modules",
        #row_km = 4, 
        column_km = 2,
        row_split = paste0("clstr", pa$clustering),
        row_gap = unit(5, "mm"),
        column_gap = unit(5, "mm"),
        # grid.text(matrix(textMatrix)),
        # border = TRUE,
        border = TRUE,
        col = col_fun,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1f", d7.SECONDtreatment_text[i, j]), x, y, gp = gpar(fontsize = 10))
        })
dev.off()
```






CASE combined
```{r}
sizeGrWindow(10,10)
# Will display correlations and their p-values
CASE.matrix <-  paste(signif(moduleTraitCor_CASE, 2), "\n(",
                                    signif(moduleTraitPvalue_CASE, 3), ")", sep = "")
par(mar = c(8, 9.5, 5, 3));
# Display the correlation values within a heatmap plot
png("outputs/CASE_heatmap.png", 800, 800, pointsize=20)
labeledHeatmap(Matrix = moduleTraitCor_CASE,
               xLabels = names(Traits.CASE),
               yLabels = names(MEs),
               ySymbols = names(MEs),
               colorLabels = TRUE,
               colors = blueWhiteRed(50),
               textMatrix = CASE.matrix,
               setStdMargins = FALSE,
               cex.text = 0.8,
               zlim = c(-0.6,0.6),
               main = paste("Module-trait relationships - CASE"))
dev.off()
```

```{r}
# this heatmap looks better
d7.GROUPtreatment_text<-  as.matrix(signif(moduleTraitPvalue_Group, 4))
d7.GROUPtreatment_cor <-  as.matrix(signif(moduleTraitCor_Group, 4))
pa = cluster::pam(d7.GROUPtreatment_cor, k = 3)
col_fun = colorRamp2(c(-0.5, 0, 0.5), c("blue", "white", "red"))

pdf("/home/mguidry/wgcna_exercise/Day7_Group_heatmap.pdf", width=5, height=6)
Heatmap(moduleTraitCor_Group, 
        name = "gene_cor", 
        rect_gp = gpar(col = "grey", lwd = 1),
        column_title = "Day 7 WGCNA - Treatment Groups", 
        column_title_gp = gpar(fontsize = 12, fontface = "bold"),
        # row_title = "WGCNA modules",
        #row_km = 4, 
        column_km = 3,
        row_split = paste0("clstr", pa$clustering),
        row_gap = unit(5, "mm"),
        column_gap = unit(5, "mm"),
        # grid.text(matrix(textMatrix)),
        # border = TRUE,
        border = TRUE,
        col = col_fun,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1f", d7.GROUPtreatment_text[i, j]), x, y, gp = gpar(fontsize = 10))
        })
dev.off()
```


Module eigengene -  MEs boxplots by treatment group
```{r}
# MEs_table             <- MEs # new table for plotting
# MEs_table$sample <- rownames(MEs) # call rows as coolumn to merge with treatment data
# MEsPlotting           <- merge(treatment_data, MEs_table, by = 'sample') # merge
# MEsPlotting2           <- MEsPlotting[,-5] # ommit the library column
# MEsPlotting3           <- MEsPlotting2[,-5] # ommit the extraction column
# MEsPlotting_melt      <- melt(MEsPlotting3, id=c('sample', 'effluent', 'pCO2'))
# 
# #plot it
# png("outputs/ME_Boxplot.png", 600, 1000, pointsize=20)
# ggplot(MEsPlotting_melt, aes(x=effluent, y=value, fill = factor(pCO2), shape=pCO2)) +
#           geom_boxplot(aes(middle = mean(value)), position=position_dodge(0.8), outlier.size = 0, alpha = 0.5) +
#           stat_summary(fun = mean, color = "black", position = position_dodge(0.75),
#                        geom = "point", shape = 19, size = 3,
#                        show.legend = FALSE) +
#           ylab("ModuleEigengene") +
#           ylim(-0.5,0.5) +
#           scale_fill_manual(values=c("#56B4E9","#D55E00")) +
#           geom_hline(yintercept=0, linetype='dotted', col = 'black', size = 1)+
#           theme_bw() +
#           theme(legend.position = "none")
#           facet_wrap(~variable)
# dev.off()
```
